src/main/scala/PopModel.scala

```diff
@@ -18,131 +18,203 @@
 
-  def calc (
-    modelName: Option[String] = None,
-    eventNames: List[String],
+  def calc(
+    modelName: String,
+    eventNames: Seq[String],
     appName: String,
     duration: Int = 0,
-    offsetDate: Option[String] = None)(implicit sc: SparkContext): Option[RDD[(String, Float)]] = {
+    offsetDate: Option[String] = None)(implicit sc: SparkContext): RDD[(ItemID, Float)] = {
 
     // todo: make end manditory and fill it with "now" upstream if not specified, will simplify logic here
     // end should always be 'now' except in unusual conditions like for testing
-    val end = if (offsetDate.isEmpty ) DateTime.now else {
+    val end = if (offsetDate.isEmpty) DateTime.now else {
       try {
         ISODateTimeFormat.dateTimeParser().parseDateTime(offsetDate.get)
       } catch {
-        case e: IllegalArgumentException => e
+        case e: IllegalArgumentException =>
           logger.warn("Bad end for popModel: " + offsetDate.get + " using 'now'")
           DateTime.now
       }
     }
 
+    val interval = new Interval(end.minusSeconds(duration), end)
+
     // based on type of popularity model return a set of (item-id, ranking-number) for all items
-    logger.info(s"PopModel ${modelName} using end: ${end}, and duration: ${duration} ")
+    logger.info(s"PopModel $modelName using end: $end, and duration: $duration ")
+
+    // if None? debatable, this is either an error or may need to default to popular, why call popModel otherwise
     modelName match {
-      case Some("popular") => calcPopular(appName, eventNames, new Interval(end.minusSeconds(duration), end))
-      case Some("trending") => calcTrending(appName, eventNames, new Interval(end.minusSeconds(duration), end))
-      case Some("hot") => calcHot(appName, eventNames, new Interval(end.minusSeconds(duration), end))
-      case _ => None // debatable, this is either an error or may need to default to popular, why call popModel otherwise
+      case BackfillType.Popular     => calcPopular(appName, eventNames, interval)
+      case BackfillType.Trending    => calcTrending(appName, eventNames, interval)
+      case BackfillType.Hot         => calcHot(appName, eventNames, interval)
+      case BackfillType.Random      => calcRandom(appName, eventNames, interval)
+      case BackfillType.UserDefined => sc.emptyRDD
+      case unknownBackfillType =>
+        logger.warn(
+          s"""
+             |Bad backfills param type=[$unknownBackfillType] in engine definition params, possibly a bad json value.
+             |Use one of the available parameter values ($BackfillType).""".stripMargin)
+        sc.emptyRDD
     }
+
+  }
+
+  /** Create random rank for all items */
+  def calcRandom(
+    appName: String,
+    eventNames: Seq[String],
+    interval: Interval)(implicit sc: SparkContext): RDD[(ItemID, Float)] = {
+
+    val events = eventsRDD(appName, eventNames, interval)
+    val itemRDD1 = events.map(_.targetEntityId).filter(_.isDefined).map(_.get).distinct()
```

Comment:
random rank and user defined must be applied to every item whether they have events or not.